File: zenwave-jpa-example/src/test/java/io/zenwave360/example/core/implementation/CustomerServiceTest.java
package io.zenwave360.example.core.implementation;

import io.zenwave360.example.config.ServicesInMemoryConfig;
import io.zenwave360.example.core.domain.Customer;
import io.zenwave360.example.core.inbound.dtos.CustomerSearchCriteria;
import io.zenwave360.example.infrastructure.jpa.inmemory.CustomerRepositoryInMemory;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.data.domain.PageRequest;

import static org.junit.jupiter.api.Assertions.*;

/** Acceptance Test for CustomerService. */
class CustomerServiceTest {

    private final Logger log = LoggerFactory.getLogger(getClass());

    ServicesInMemoryConfig context = new ServicesInMemoryConfig();

    CustomerServiceImpl customerService = context.customerService();

    CustomerRepositoryInMemory customerRepository = context.customerRepository();

    @BeforeEach
    void setUp() {
        context.reloadTestData();
    }

    @Test
    void createCustomerTest() {
        var input = new Customer();
        // TODO fill input data
        // input.setName("");
        // input.setEmail("");
        // input.setAddresses(List.of(new Address()));
        var customer = customerService.createCustomer(input);
        assertNotNull(customer.getId());
        assertTrue(customerRepository.containsEntity(customer));
    }

    @Test
    void getCustomerTest() {
        var id = 1L; // TODO fill id
        var customer = customerService.getCustomer(id);
        assertTrue(customer.isPresent());
    }

    @Test
    void updateCustomerTest() {
        var id = 1L; // TODO fill id
        var input = new Customer();
        // TODO fill input data
        // input.setName("");
        // input.setEmail("");
        // input.setAddresses(List.of(new Address()));
        assertTrue(customerRepository.containsKey(id));
        var customer = customerService.updateCustomer(id, input);
        assertTrue(customer.isPresent());
        assertTrue(customerRepository.containsEntity(customer.get()));
    }

    @Test
    void deleteCustomerTest() {
        var id = 1L; // TODO fill id
        assertTrue(customerRepository.containsKey(id));
        customerService.deleteCustomer(id);
        assertFalse(customerRepository.containsKey(id));
    }

    @Test
    void searchCustomersTest() {
        var searchCriteria = new CustomerSearchCriteria();
        var results = customerService.searchCustomers(searchCriteria, PageRequest.of(0, 10));
        Assertions.assertNotNull(results);
        Assertions.assertFalse(results.isEmpty());
    }

}

File: examples/kustomer-address-jpa/src/test/kotlin/io/zenwave360/examples/kotlin/service/impl/CustomerServiceTest.kt
package io.zenwave360.examples.kotlin.service.impl

import io.zenwave360.examples.kotlin.config.*
import io.zenwave360.examples.kotlin.domain.*
import io.zenwave360.examples.kotlin.service.*
import io.zenwave360.examples.kotlin.service.dtos.*
import io.zenwave360.examples.kotlin.service.impl.mappers.*
import io.zenwave360.examples.kotlin.repository.jpa.*
import io.zenwave360.examples.kotlin.repository.jpa.inmemory.*

import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mapstruct.factory.Mappers
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.PageRequest

import java.util.Map
import java.util.Optional
import java.time.*
import java.math.BigDecimal

import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Assertions.assertFalse
import org.mockito.Mockito.*

/**
 * Acceptance Test for CustomerService.
 */
class  CustomerServiceTest {

    private val log: Logger = LoggerFactory.getLogger(javaClass)

    val context = ServicesInMemoryConfig()
    val customerService: CustomerService = context.customerService()

    val customerRepository: CustomerRepositoryInMemory = context.customerRepository()


    @BeforeEach
    fun setUp() {
        context.reloadTestData()
    }


    @Test
    fun createCustomerTest() {
        val input = Customer()
        input.name = "name"
        input.email = "me@email.com"
        input.addresses = mutableListOf(Address().apply {
            street = "street"
            city = "city"
        })
        val customer = customerService.createCustomer(input)
        assertNotNull(customer.id)
        assertTrue(customerRepository.containsEntity(customer))
}

    @Test
    fun getCustomerTest() {
        val id: Long = 1L
        val customer = customerService.getCustomer(id)
        assertTrue(customer.isPresent)
}

    @Test
    fun updateCustomerTest() {
        val id: Long = 1L
        val input = Customer()
        input.name = "name"
        input.email = "me@email.com"
        input.addresses = mutableListOf(Address().apply {
            street = "street"
            city = "city"
        })
        assertTrue(customerRepository.containsKey(id))
        val customer = customerService.updateCustomer(id, input)
        assertTrue(customer.isPresent)
        assertTrue(customerRepository.containsEntity(customer.get()))
}

    @Test
    fun deleteCustomerTest() {
        val id: Long = 1L
        assertTrue(customerRepository.containsKey(id))
        customerService.deleteCustomer(id)
        assertFalse(customerRepository.containsKey(id))
}

    @Test
    fun searchCustomersTest() {
        val searchCriteria = CustomerSearchCriteria()
        val results = customerService.searchCustomers(searchCriteria, PageRequest.of(0, 10))
        assertNotNull(results)
        assertFalse(results.isEmpty)
}

}
